### ISSUE TITLE
mypy suggests `Iterator[Never]`, wants `Iterator[None]` as return type of contextmanager

### ISSUE THREAD

**Issue by alexei** [python/mypy#18086]
A superficial take on a context manager (https://mypy-play.net/?mypy=latest&python=3.12&gist=68fae9f1b14467dc3a586f2473c9c419):

```python
from contextlib import contextmanager


@contextmanager
def foo() -> None:
    yield
```

results in:

```
main.py:4: error: Argument 1 to "contextmanager" has incompatible type "Callable[[], None]"; expected "Callable[[], Iterator[Never]]"  [arg-type]
main.py:5: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
```

---

If I specify `Iterator[Never]` as suggested, it still doesn't work (https://mypy-play.net/?mypy=latest&python=3.12&gist=d644d5c880c5f04b0c54941a7cc8c4b2):

```python
from contextlib import contextmanager
from typing import Iterator, Never


@contextmanager
def foo() -> Iterator[Never]:
    yield
```

results in:

```
main.py:7: error: Yield value expected  [misc]
```

---

Inspired by https://github.com/python/mypy/issues/3551 I switched it to `None` (https://mypy-play.net/?mypy=latest&python=3.12&gist=e825ee207b3fbcad373163f5618f467e):

```python
from contextlib import contextmanager
from typing import Iterator, Never


@contextmanager
def foo() -> Iterator[None]:
    yield
```

And now it works:

```
Success: no issues found in 1 source file
```

How? Wasn't `Never` (as in "never returns") supposed to not ask for a return value?

**Comment by sterliakov** [comment]
> Wasn't Never (as in "never returns") supposed to not ask for a return value?

Nope, `Iterator[Never]` is something that yields values of type `Never` - essentially never yields at all. Your function does yield `None` in fact, bare `yield` and bare `return` have "implied" `None`, and `None` is not assignable to `Never`. 

"yield value expected" means that your function should yield non-None but uses bare `yield`. 

`Never` in the first error message might be misleading, but that comes from solving a typevar in `contextmanager` signature. That typevar can't be solved to `None` because your original signature does not return an `Iterable`.

In my understanding `mypy` behaviour is correct here.

**Comment by A5rocks** [comment]
Perhaps when mypy emits this error:
```
main.py:5: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
```
it should replace the return type with `Generator[AnyType.from_error]`?

**Comment by alexei** [comment]
@sterliakov thanks for the explanation; it makes sense

@A5rocks yeah, it might be better to suggest `Generator[Any]` instead of `Iterator[Never]`
